src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+{
+    bool certificationDeclarationExists = false;
+    bool attestationNonceExists         = false;
+    bool timestampExists                = false;
+    bool firmwareInfoExists             = false;
+    size_t vendorReservedIdx            = 0;
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    while ((error = tlvReader.Next()) == CHIP_NO_ERROR)
+    {
+        uint64_t tag = tlvReader.GetTag();
+
+        if (TLV::IsContextTag(tag))
+        {
+            switch (TLV::TagNumFromTag(tag))
+            {
+            case kCertificationDeclarationTagId:
+                VerifyOrReturnError(certificationDeclarationExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(certificationDeclaration));
+                certificationDeclarationExists = true;
+                break;
+            case kAttestationNonceTagId:
+                VerifyOrReturnError(attestationNonceExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(attestationNonce));
+                attestationNonceExists = true;
+                break;
+            case kTimestampTagId:
+                VerifyOrReturnError(timestampExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(timestamp));
+                timestampExists = true;
+                break;
+            case kFirmwareInfoTagId:
+                VerifyOrReturnError(firmwareInfoExists == false, CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+                ReturnErrorOnFailure(tlvReader.Get(firmwareInfo));
+                firmwareInfoExists = true;
+                break;
+            default:
+                return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT;
+            }
+        }
+        else if (TLV::IsProfileTag(tag))
+        {
+            // vendor fields
+            bool seenProfile = false;
+            uint16_t currentVendorId;
+            uint16_t currentProfileNum;
+
+            currentVendorId   = TLV::VendorIdFromTag(tag);
+            currentProfileNum = TLV::ProfileNumFromTag(tag);
+            if (!seenProfile)
+            {
+                seenProfile = true;
+                vendorId    = currentVendorId;
+                profileNum  = currentProfileNum;
+            }
+            else
+            {
+                // check that vendorId and profileNum match in every Vendor Reserved entry
+                VerifyOrReturnError(currentVendorId == vendorId && currentProfileNum == profileNum,
+                                    CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
```

Comment:
This is not mandatory according to spec.

Reply:
ok, thanks

Reply:
We cannot assume anything: there may be duplicate vendorId/profileNum. Eventually each entry should be accessible, even if duplicated. That is why I suggested we eventually (step 2) move to have this be a class, and have an iteration API for the vendor tags after parsing.

Reply:
So duplicate tags can happen on deconstruction.   

Should we cause an error on "add vendor specific data" for creating?

Reply:
careful reading of the spec indicates there's nothing written having duplicate tags.